src/mlpack/tests/feedforward_network_test.cpp

```diff
@@ -189,6 +189,65 @@ TEST_CASE("CheckCopyMovingReparametrizationNetworkTest", "[FeedForwardNetworkTes
   CheckMoveFunction<>(model1, trainData, trainLabels, 1);
 }
 
+/**
+ * Check whether copying and moving network with select layer is working or not.
+ */
+TEST_CASE("CheckCopyMovingSelectNetworkTest", "[FeedForwardNetworkTest]")
+{
+  // Load the dataset.
+  arma::mat trainData;
+  data::Load("thyroid_train.csv", trainData, true);
+
+  arma::mat trainLabels = trainData.row(trainData.n_rows - 1);
+  trainData.shed_row(trainData.n_rows - 1);
+  arma::mat SelectedLabel = trainLabels.submat(0,0,trainData.n_rows/2,0);
+  trainLabels=SelectedLabel;
```

Comment:
I think that these lines may be the build problem.  I don't understand what the goals of the line are... can you explain?

Reply:
Hi @rcurtin , Here I am selecting the submatrix containing the first column and half of the rows to verify the working of select layer(selecting a particular column). Also the tests were passing locally.

Reply:
If you didn't compile with `-DDEBUG=ON`, my guess is that bounds checks were not being performed.

I think the `submat()` call is wrong here; wouldn't the last parameter need to be `trainData.n_cols - 1`, not `0`?  You might want to check the Armadillo documentation.

Reply:
Armadillo documentation says : **X.submat( first_row, first_col, last_row, last_col )** and in Select::Forward() function the first column and last column are passed the same value(index) hence one column. Basically we are selecting a particular column from the matric using select layer. Please Correct me if I am wrong.

Reply:
Okay, thanks for the explanation.  If you're going to only select one column for the labels, shouldn't you also select only one column for the data?  Also, it may be clearer to select only one column as just `trainLabels = trainLabels.col(0);`. :)

When I ran this locally after configuring CMake with `-DDEBUG=ON`, I saw that there was an invalid matrix size used somewhere in the code, so I guessed it was here, but I think you are right that you are using `submat()` correctly.  I didn't dig in with `gdb` to figure out exactly where the exception was coming from.  If you build locally (with `-DDEBUG=ON`), you can probably reproduce the issue, and then perhaps dig in more with a debugger, or reason about the code to fix it---the issue might be as simple as the training data not having the same size as the training labels. :+1:

Reply:
Thanks for review @rcurtin , I had same thing in my mind there should be only one train label but then number of training columns and Train labels will not match before passing into the model because initially we are passing all training columns and then by select layer it is selecting one column. I agree that the issue is simple as once I am able to build without error, it will definitely be solved. But When i tried to check this , I am not able to build mlpack(master branch) locally due to the following error in make:
<pre><font color="#729FCF"><b>mlpack-master/build</b></font>$ make
[  0%] <font color="#729FCF"><b>Updating arma_config.hpp (if necessary)</b></font>
[  0%] Built target mlpack_arma_config
<font color="#729FCF"><b>Moving header files to include/mlpack/</b></font>
/bin/sh: 1: Syntax error: &quot;(&quot; unexpected
src/mlpack/CMakeFiles/mlpack_headers.dir/build.make:57: recipe for target &apos;mlpack_headers&apos; failed
make[2]: *** [mlpack_headers] Error 2
CMakeFiles/Makefile2:208: recipe for target &apos;src/mlpack/CMakeFiles/mlpack_headers.dir/all&apos; failed
make[1]: *** [src/mlpack/CMakeFiles/mlpack_headers.dir/all] Error 2
Makefile:140: recipe for target &apos;all&apos; failed
make: *** [all] Error 2
</pre>

Reply:
That sounds unrelated to the code here.  Maybe you should clean/remove your build directory and try configuring/building again.  (Out of curiosity, is there a `(` in your directory name or something like that?)

Reply:
Surprisingly it was the same issue you pointed out regarding the `(` in directory name. So i was able to build properly. But now when i am doing `make mlpack_test` it is crashing even after changing number of cores. Also there are warnings in ann_layer_test.cpp.o and is crashing here only.

Reply:
Did you try debugging it with a debugger to see where it crashes?  "It crashes" is a starting point, not a final report.  If you dig in to where it crashes, the error should become obvious, and then you can fix it. :+1:

Reply:
Sorry, I misread what you wrote.  You are saying instead it is a compilation failure.  In this case, you should be able to use the output from the compiler to figure out what is wrong and then fix the issue.  Then, if the test fails when you run it, you can use a debugger like I suggested above. :)

Reply:
Actually now I have tried `make mlpack_test -d` and I am not getting any proper error. The system is freezing at ann_layer_tests.cpp.o while I have nothing to do with it.

Reply:
What have you tried to debug it?  Regardless of whether or not your code (or you) have anything to do with it, you'll evidently still need to solve the problem in order to fix the original problem in this PR.

It sounds like you are running out of RAM.  Try building with only one core or closing other applications that may be using RAM.

Reply:
Thanks @rcurtin ,I have successfully done the compiling. Can u suggest debugging techniques while testing so that I will do the debugging properly .I have tried `--help` for testing but didn't find anything except -b for breaking into debugger but  its not useful. I have changed the syntax of matrices and avoided use of `arma::submat` as discussed with you to select one training label. Now I am receiving 
<pre>due to unexpected exception with message:
  Mat::submat(): indices out of bounds or incorrectly used
</pre>
and for other layers in feedforward_network_test.cpp i am unexpextedly getting
<pre>due to unexpected exception with message:
  Log::Assert() failed: Target class out of range.
</pre>

Reply:
Awesome, glad you got it compiled.  For this debugging, I would suggest using a debugger like GDB.  Here's a reasonable tutorial on its usage: https://www.cs.cmu.edu/~gilpin/tutorial/  (there are lots of other tutorials too, and there may be other good C++ debuggers too, but gdb is the one that I use, at least).

I think what you want to do is set a breakpoint when the exception from `submat()` is thrown.  You could do this, once you have the program loaded, by typing `catch throw` at the gdb prompt so that it will break when the exception is thrown.  Then, you can run your test, and once the exception is thrown, you can use `bt` to look at the backtrace and figure out what is invalid about the `submat()` call (and exactly where it's being called).  I've omitted a lot of steps there, but between the tutorial I linked to and the gdb documentation, I think that should be sufficient to get things figured out. :+1:

Reply:
@rcurtin.Thanks for getting me familiar with gdb.I followed the procedure you told. One thing which i have in my mind was about the library path which i am using is :<pre>LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu
</pre>
but in the documentation the mentioned path is <pre>LD_LIBRARY_PATH=/usr/local/lib
</pre> which is giving the following error while testing  <pre>bin/mlpack_test: symbol lookup error: bin/mlpack_test: undefined symbol: arma_H5T_NATIVE_ULONG
</pre> so i changed my library path to the formal one. Is this library path correct? as maybe some dependicies are missing in this path as unexpected errors are coming in so many tests. I think issue is with the dependencies. By the time i have pushed the test case which i think is logically correct. 
Gdb output on my pc is as shown below
<pre>Using host libthread_db library &quot;/lib/x86_64-linux-gnu/libthread_db.so.1&quot;.
mlpack version: mlpack 3.4.3
armadillo version: 8.400.0 (Entropy Bandit)
<font color="#FCE94F"><b>Filters: [feedforwardnetworkTEST]</b></font>

<font color="#FCE94F"><b>Catchpoint 1 (exception thrown), 0x00007ffff5cd2d1d in __cxa_throw ()</b></font>
<font color="#FCE94F"><b>   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6</b></font>
<font color="#FCE94F"><b>(gdb) bt</b></font>
<font color="#FCE94F"><b>#0  0x00007ffff5cd2d1d in __cxa_throw ()</b></font>
<font color="#FCE94F"><b>   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6</b></font>
<font color="#FCE94F"><b>#1  0x00007ffff72320de in mlpack::Log::Assert (condition=false, </b></font>
<font color="#FCE94F"><b>    message=&quot;Target class out of range.&quot;)</b></font>
<font color="#FCE94F"><b>    at /home/anmol/mlpack april/mlpack-master/mlpack-master/src/mlpack/core/util/log.cpp:34</b></font>
<font color="#FCE94F"><b>#2  0x0000555558f485a1 in mlpack::ann::NegativeLogLikelihood&lt;arma::Mat&lt;double&gt;, arma::Mat&lt;double&gt; &gt;::Forward&lt;arma::Mat&lt;double&gt;, arma::subview&lt;double&gt; &gt; (</b></font>
<font color="#FCE94F"><b>    this=0x55555ba117a0, prediction=..., target=...)</b></font>
<font color="#FCE94F"><b>    at /home/anmol/mlpack april/mlpack-master/mlpack-master/src/mlpack/methods/ann/loss_functions/negative_log_likelihood_impl.hpp:38</b></font>
<font color="#FCE94F"><b>#3  0x0000555558f32340 in mlpack::ann::FFN&lt;mlpack::ann::NegativeLogLikelihood&lt;arma::Mat&lt;double&gt;, arma::Mat&lt;double&gt; &gt;, mlpack::ann::RandomInitialization&gt;::EvaluateWithGradient&lt;arma::Mat&lt;double&gt; &gt; (this=0x55555ba117a0, begin=0, </b></font>
<font color="#FCE94F"><b>    gradient=..., batchSize=32)</b></font>
<font color="#FCE94F"><b>    at /home/anmol/mlpack april/mlpack-master/mlpack-master/src/mlpack/methods/ann/ffn_impl.hpp:384</b></font>
<font color="#FCE94F"><b>#4  0x0000555558fbf21d in ens::AddSeparableEvaluateWithGradient&lt;mlpack::ann::FFN&lt;mlpack::ann::NegativeLogLikelihood&lt;arma::Mat&lt;double&gt;, arma::Mat&lt;double&gt; &gt;, mlpack::ann::RandomInitialization&gt;, arma::Mat&lt;double&gt;, arma::Mat&lt;double&gt;, true, true&gt;::EvaluateWithGradient (this=0x55555ba117a0, coordinates=..., begin=0, </b></font>
<font color="#FCE94F"><b>    gradient=..., batchSize=32)</b></font>
<font color="#FCE94F"><b>    at /usr/local/include/ensmallen_bits/function/add_separable_evaluate_with_gr---Type &lt;return&gt; to continue, or q &lt;return&gt; to quit---</b></font>
</pre>

Reply:
@Anmol2001 nice, glad you got gdb working!  I think that you shouldn't need to set `LD_LIBRARY_PATH` at all.  I'm assuming that you are running this from your build directory?  The documentation only mentions setting `LD_LIBRARY_PATH` as something that is necessary if you've installed mlpack to `/usr/local/`.

I would say here that you want to look further up the backtrace, to find the actual test case that is failing.  Are the labels between `0` and `numClasses - 1`?  That is what they should be.  You could even step into frame 2 (`frame 2`) and inspect the variables with `inspect` to see what the values actually are.

I'm not sure how out of sync this branch is with master, but it could also be helpful to merge master in.  Some months ago we changed how `NegativeLogLikelihood` worked, and that *could* be what's going on here (but I think it's unlikely---more likely to me seems to be invalid class labels in the network).